### DISCUSSION TITLE
Offer UTF-8 and UTF-16 text APIs on Windows?

### DISCUSSION THREAD

**Discussion by ghost** [libui-ng/libui-ng#144] [Ideas]
`uiWindowsWindowText` silently converts the UTF-16 output to UTF-8, and `uiWindowsSetWindowText` silently converts the UTF-8 input to UTF-16. But, what if I do want the original UTF-16 text of a window, or I want to set the text of a window with a UTF-16 string?

Maybe we could offer a `uiWindowsText` struct that's basically a tagged union and can store either UTF-8 or UTF-16 text, and we could have `uiWindowsTextUTF8` and uiWindowsTextUTF16` that 'extract' the UTF-8 or UTF-16 text at runtime, performing a conversion only if necessary?

**Comment by cody271** [comment]
> Maybe we could offer a `uiWindowsText` struct that's basically a tagged union and can store either UTF-8 or UTF-16 text, and we could have `uiWindowsTextUTF8` and uiWindowsTextUTF16` that 'extract' the UTF-8 or UTF-16 text at runtime, performing a conversion only if necessary?
>
This sure sounds a lot like the status quo on Windows itself already, where nearly every function has both `FooBarA()` and `FooBarW()` variants. Something I very much would like to avoid! A _lot_ more complexity for no clear benefit.. What is the specific use case here you had in mind?

**Reply by ghost** [reply]
> This sure sounds a lot like the status quo on Windows itself already, where nearly every function has both `FooBarA()` and `FooBarW()` variants. Something I very much would like to avoid!

I agree with you here! Having A and W versions of every function would be a nightmare to maintain. I'm thinking something more along the lines of this:

```c
enum uiWindowsTextKind {
    uiWindowsTextKindUTF8,
    uiWindowsTextKindUTF16,
};

struct uiWindowsText {
    uiWindowsTextKind kind;
    void *inner;
};

char *uiWindowsTextUTF8(const uiWindowsText *t)
{
    switch (t->kind) {
    case uiWindowsTextKindUTF8:
        return (char *)t->data;
    case uiWindowsTextKindUTF16:
        // Do conversion...
    }
}

// And vice-versa for `uiWindowsTextUTF8`...
```

We don't need A and W versions of each function if we do it like this. For example:

```c
void uiWindowsSetWindowText(HWND hwnd, const uiWindowsText *text)
{
    setWindowText(hwnd, uiWindowsTextUTF16(text));
}
```

We would need to work out allocation and de-allocation though.

> What is the specific use case here you had in mind?

I don't really have a specific use case; I just noticed that *libui-ng* silently and unconditionally (even on very large text buffers!) converts between UTF-8 and UTF-16. I think we could offer a little more control in the API.

Also, on a slightly related note, it might be a good idea to memoize some text-related functions. For example, if I query the current text of a window and the text content hasn't changed since I last queried it, then *libui-ng* shouldn't have to convert the text from UTF-16 to UTF-8 again.

**Comment by cody271** [comment]
Well I'm certainly open to a PR here if reasonable and high quality, but I would really advise you against this entire line of thinking to be honest.. We have much higher priority things to be improving IMHO.

None of these functions will ever be called in a hot enough code path to justify these sorts of optimizations you're proposing.

The only place I can see right now where this might be worth the effort is: offering a UTF-16 version of `uiAttributedString` API, for text rendering inside custom drawn `uiArea` based content. That is the only case AFAIK where the conversions overhead happens >1FPS.

```
Premature Optimization Is the Root of All Evil.

- "The Art of Computer Programming" by Prof. Donald Knuth
```

**Reply by ghost** [reply]
I see your point; this is definitely low priority even among other low priority issues. The idea came to mind, so I was wondering if this had been discussed before. Even if this is not implemented in the near future (or at all!), I'd like to put it on the table.

I would agree that these optimizations would offer relatively minor performance benefits, though memoization, for example, probably wouldn't hinder performance even if the data being memoized is constantly changing. Memoization doesn't really make much sense for small data (e.g., integers, booleans, etc.) though it could help to prevent unnecessary, duplicate buffer allocations.

But yes, this topic is probably not worth being discussed much further at the moment as there are more pressing issues.

**Reply by szanni** [reply]
> None of these functions will ever be called in a hot enough code path to justify these sorts of optimizations you're proposing.

Exactly this. We are converting tiny amounts of text, ease of API should be top priority in my opinion. I am pretty certain benchmarks would prove the performance overhead to be negligible (happy to be proven wrong). We are processing a few UI elements, not a database with millions of entries.

> The only place I can see right now where this might be worth the effort is: offering a UTF-16 version of `uiAttributedString` API, for text rendering inside custom drawn `uiArea` based content. That is the only case AFAIK where the conversions overhead happens >1FPS.

These were my thoughts too. This _might_ make sense for text editors displaying huge amounts of text. Again I would be curious about what a benchmark would show.
This would make sense from an API perspective too. The `uiAttributedString` API is already quite complex. If you really need UTF16 for performance reasons, you are already dealing with an entirely different set of problems handling your data. Needing to use the more complex `uiAttributedString` API should not make any difference at that point.
In the end you probably need to convert every time you load/store your data into a different encoding. I am honestly not sure how much would be gained by any of this.

> But yes, this topic is probably not worth being discussed much further at the moment as there are more pressing issues.

Sadly very much so.
